brickstrap.sh

```diff
@@ -120,96 +146,172 @@ function help() {
+                    BR_COMPONENTS="'$OPTARG'"
+                else
+                    BR_COMPONENTS="$BR_COMPONENTS '$OPTARG'"
+                fi
+            ;;
+            d) ROOTDIR="$OPTARG";;
+            \?) # unknown flag
+                brp_help "Invalid option specified"
                 exit 1
             ;;
-        esac
-    done
-
-    shift $((OPTIND-1))
-
-    if [ "${cmd}" == "run-hook" ]; then
-      run_hook_arg=$1
-    fi
+            \:) # missing argument
+                brp_help "Missing argument for option: -$BRP_OPT"
+                exit 1
+            ;;
+            esac
+        done
 
-    cmd=${cmd:=$1}
+        shift $((OPTIND-1))
 
-    shift 1
-    OPTIND=1
-done
+        if [ "$BRP_CMD" == "run-hook" ] || [ "$BRP_CMD" == "shell" ]; then
+            BRP_CMD_ARG="$1"
+        fi
 
-[ "${cmd}" = "" ] && help && exit 1
-[ "${cmd}" = "run-hook" ] && [ "${run_hook_arg}" = "" ] && help && exit 1
-debug "cmd: ${cmd}"
+        BRP_CMD="${BRP_CMD:=$1}"
 
-#####################################################################
-### Set up the variables for the commands
+        shift 1
+        OPTIND=1
+    done
+}
 
-SCRIPT_PATH=$(dirname $(readlink -f "$0"))
+function brp_sanity_check_cli_options()
+{
+    if [ $# -eq 0 -o -z "$1" ]; then
+        brp_help 'No command specified'
+        exit 1
+    elif [ "$1" = "run-hook" ] && [ $# -eq 1 -o -z "$2" ]; then
+        brp_help 'No hook specified'
+        exit 1
+    fi
+    debug "cmd: $1"
+}
 
-debug "SCRIPT_PATH: ${SCRIPT_PATH}"
+function brp_validate_cli_options()
+{
+    brp_sanity_check_cli_options "$@" && \
+    brp_validate_project_name && \
+    brp_validate_component_names
+}
 
-export DEBIAN_FRONTEND=noninteractive
-export DEBCONF_NONINTERACTIVE_SEEN=true
-export LC_ALL=C LANGUAGE=C LANG=C
-export PATH=$PATH:/usr/local/sbin:/usr/sbin:/sbin
 
-[ -r "${BOARDDIR}" ] || BOARDDIR="${SCRIPT_PATH}/${BOARDDIR}"
-[ -r "${BOARDDIR}" ] || fail "cannot find target directory: ${BOARDDIR}"
-[ -r "${BOARDDIR}/multistrap.conf" ] \
-    || fail "cannot read multistrap config: ${BOARDDIR}/multistrap.conf"
+#####################################################################
+### Set up the variables for the commands
 
-BOARDDIR=$(readlink -f "${BOARDDIR}")
+function brp_init_env()
+{
 
-for SYSTEM_KERNEL_IMAGE in /boot/vmlinuz-*; do
-    [ -r "${SYSTEM_KERNEL_IMAGE}" ] \
-        || fail "Cannot read ${SYSTEM_KERNEL_IMAGE} needed by guestfish." \
-        "Set permission with 'sudo chmod +r /boot/vmlinuz-*'."
-done
+    debug "SCRIPT_PATH: $(br_script_path)"
 
-if [[ "$(sysctl -ne kernel.unprivileged_userns_clone)" == "0" ]]; then
-    fail "Unprivileged user namespace clone is disabled. Enable it by running" \
-        "'sudo sysctl -w kernel.unprivileged_userns_clone=1'."
-fi
+    export DEBIAN_FRONTEND=noninteractive
+    export DEBCONF_NONINTERACTIVE_SEEN=true
+    export LC_ALL=C LANGUAGE=C LANG=C
+    export PATH=$PATH:/usr/local/sbin:/usr/sbin:/sbin
```

Comment:
Yes. Also, yes: let's save that for a follow up PR because the changeset is big enough as it its. ;)